File: src/server/storage/store.ts
import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { promisify } from 'util';
import { MAILDOMAINPREFIX } from '../../config/config';

const MAILBOX_DIR = path.join(__dirname, 'maildir');
const mkdir = promisify(fs.mkdir);
const writeFile = promisify(fs.writeFile);

async function ensureMaildirStructure(userDir: string) {
	for (const sub of ['tmp', 'new', 'cur']) {
		const dir = path.join(userDir, sub);
		if (!fs.existsSync(dir)) {
			await mkdir(dir, { recursive: true });
		}
	}
}

function generateMailFilename(): string {
	const timestamp = Date.now();
	const random = crypto.randomBytes(6).toString('hex');
	const pid = process.pid;
	const hostname = require('os').hostname();
	return `${timestamp}.${pid}.${hostname}.${random}`;
}

export async function saveEmail(recipient: string, data: string) {
	const localPart = recipient.split(MAILDOMAINPREFIX)[0];
	const userDir = path.join(MAILBOX_DIR, localPart);

	await ensureMaildirStructure(userDir);

	const fileName = generateMailFilename();
	const tmpPath = path.join(userDir, 'tmp', fileName);
	const newPath = path.join(userDir, 'new', fileName);

	await writeFile(tmpPath, data, 'utf-8');

	fs.renameSync(tmpPath, newPath);

	console.log(`Saved email to ${newPath}`);
}

File: smtp/store.go
package smtp

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const MAILBOX_DIR = "maildir"
const MAILDOMAIN_PREFIX = "~"

func ensureMaildirStructure(userDir string) error {
	for _, sub := range []string{"tmp", "new", "cur"} {
		dir := filepath.Join(userDir, sub)
		if _, err := os.Stat(dir); os.IsNotExist(err) {
			if err := os.MkdirAll(dir, 0755); err != nil {
				return err
			}
		}
	}
	return nil
}

func generateMailFilename() (string, error) {
	timestamp := time.Now().UnixNano()
	randomBytes := make([]byte, 6)
	_, err := rand.Read(randomBytes)
	if err != nil {
		return "", err
	}
	randomHex := hex.EncodeToString(randomBytes)
	pid := os.Getpid()
	hostname, _ := os.Hostname()

	return fmt.Sprintf("%d.%d.%s.%s", timestamp, pid, hostname, randomHex), nil
}

func SaveEmail(recipient string, data string) error {
	parts := strings.Split(recipient, MAILDOMAIN_PREFIX)
	if len(parts) == 0 {
		return fmt.Errorf("invalid recipient format")
	}
	localPart := parts[0]
	userDir := filepath.Join(MAILBOX_DIR, localPart)

	if err := ensureMaildirStructure(userDir); err != nil {
		return err
	}

	filename, err := generateMailFilename()
	if err != nil {
		return err
	}

	tmpPath := filepath.Join(userDir, "tmp", filename)
	newPath := filepath.Join(userDir, "new", filename)

	if err := os.WriteFile(tmpPath, []byte(data), 0644); err != nil {
		return err
	}

	if err := os.Rename(tmpPath, newPath); err != nil {
		return err
	}

	fmt.Printf("Saved email to %s\n", newPath)
	return nil
}
